components/brave_wallet/browser/internal/orchard_storage/orchard_shard_tree_delegate.h

```diff
@@ -0,0 +1,160 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_INTERNAL_ORCHARD_STORAGE_ORCHARD_SHARD_TREE_DELEGATE_H_
+#define BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_INTERNAL_ORCHARD_STORAGE_ORCHARD_SHARD_TREE_DELEGATE_H_
+
+#include <utility>
+#include <vector>
+
+#include "base/memory/raw_ref.h"
+#include "brave/components/brave_wallet/common/zcash_utils.h"
+
+namespace brave_wallet {
+
+class ZCashOrchardStorage;
+
+// Leaf position of checkpoint.
+using CheckpointTreeState = std::optional<uint32_t>;
+
+// Checkpointed leafs are not pruned so they could be used
+// as anchors for building shielded transactions.
+// Last Orchard commitment in a block is used as a checkpoint.
+struct OrchardCheckpoint {
+  OrchardCheckpoint();
+  OrchardCheckpoint(CheckpointTreeState, std::vector<uint32_t>);
+  ~OrchardCheckpoint();
+  OrchardCheckpoint(const OrchardCheckpoint& other);
+  OrchardCheckpoint& operator=(const OrchardCheckpoint& other);
+  OrchardCheckpoint(OrchardCheckpoint&& other);
+  OrchardCheckpoint& operator=(OrchardCheckpoint&& other);
+
+  bool operator==(const OrchardCheckpoint& other) const = default;
+
+  // Leaf position of the checkpoint.
+  CheckpointTreeState tree_state_position;
+  // List of note positions that were spent at this checkpoint.
+  std::vector<uint32_t> marks_removed;
+};
+
+struct OrchardCheckpointBundle {
+  OrchardCheckpointBundle(uint32_t checkpoint_id, OrchardCheckpoint);
+  ~OrchardCheckpointBundle();
+  OrchardCheckpointBundle(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle& operator=(const OrchardCheckpointBundle& other);
+  OrchardCheckpointBundle(OrchardCheckpointBundle&& other);
+  OrchardCheckpointBundle& operator=(OrchardCheckpointBundle&& other);
+
+  bool operator==(const OrchardCheckpointBundle& other) const = default;
+
+  // The block height serves as the checkpoint identifier.
+  uint32_t checkpoint_id = 0;
+  OrchardCheckpoint checkpoint;
+};
+
+// Address of a subtree in the shard tree.
+struct OrchardShardAddress {
+  uint8_t level = 0;
+  uint32_t index = 0;
+
+  bool operator==(const OrchardShardAddress& other) const = default;
```

Comment:
Do you really need to define these comparison operators in these structs?

Reply:
They are required for test purposes

Reply:
Acknowledged